### PR TITLE
GitlabUrlReader to check for modified with SHA of only commits that modify filepath

### PR DESCRIPTION
**Pull request by adrianke77**
## Hey, I just made a Pull Request!

Original logic in `readTree` in GitlabUrlReader always compares the SHA of the last commit of the branch when checking if there have been modifications, but this is inaccurate for cases where a specified path within the branch is also given, as the last commit of the whole branch may not have any changes within the specified path.

This patch makes `readTree' perform the SHA comparison with the last of only commits that affect the specified path.

Only GitlabUrlReader is updated here as I'm not sure if I correctly understand what GithubUrlReader is doing, and I don't have an example of the other Git repositiory types to test with. 

#### :heavy_check_mark: Checklist

- [ :heavy_check_mark: ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ NA ] Added or updated documentation
- [  :heavy_check_mark: ] Tests for new functionality and regression tests for bug fixes
- [ NA ] Screenshots attached (for UI changes)
- [  :heavy_check_mark: ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by OrkoHunter — COMMENTED** [review]
This should be useful. One thought from me - I feel that if we can the actual 	`ETag` returned by GitLab as the etag in URL Reader instead of the commit SHA, we can avoid a little bit of complexity here. Plus this should be easy to generalize over other providers as well e.g. GitHub. What do you think?

**Comment by adrianke77** [comment]
> This should be useful. One thought from me - I feel that if we can the actual `ETag` returned by GitLab as the etag in URL Reader instead of the commit SHA, we can avoid a little bit of complexity here. Plus this should be easy to generalize over other providers as well e.g. GitHub. What do you think?

It sounds like it would work, I'm not familiar with ETag caching previously but from Gitlab docs it looks like they serve to identify when a particular REST request's response would change? 

I'd like to request we get this PR through and then we could look at a general Etag based resolver that works over other Git providers separately. In Grab we're ramping up our use of TechDocs and without this PR we'll get incorrect new doc version notices and unnecessary rebuilds, as most of our docs are on only a handful of repos that also see a lot of code commits.

**Review by OrkoHunter — APPROVED** [review]
Alright - this looks good to me.

**Comment by OrkoHunter** [comment]
Need to update vale vocab list with some words https://github.com/backstage/backstage/pull/5599/checks?check_run_id=2515257128

File: packages/backend-common/src/reading/GitlabUrlReader.ts [hunk 1]
```diff
@@ -106,25 +106,28 @@ export class GitlabUrlReader implements UrlReader {
     // ref is an empty string if no branch is set in provided url to readTree.
     const branch = ref || projectGitlabResponseJson.default_branch;
 
-    // Fetch the latest commit in the provided or default branch to compare against
-    // the provided sha.
-    const branchGitlabResponse = await fetch(
+    // Fetch the latest commit that modifies the the filepath in the provided or default branch
+    // to compare against the provided sha.
+    const branchQuery = `ref_name=${branch}`;
+    const pathQuery = !!filepath ? `&path=${filepath}` : '';
```
Comment (freben) [inline packages/backend-common/src/reading/GitlabUrlReader.ts:112]: should construct these using `URLSearchParams` instead, which also solves the proper safe quoting of values

**Review by freben — COMMENTED** [review]
Looks good, but use URLSearchParams instead.

File: packages/backend-common/src/reading/GitlabUrlReader.ts (hunk 1 above)
Comment (adrianke77) [reply packages/backend-common/src/reading/GitlabUrlReader.ts]: Updated

**Review by freben — APPROVED** [review]

**Pull request merged**